File: ios/LaneShadowTests/Sandbox/IdlePlanningVariantTests.swift
import Foundation
import LaneShadowTheme
import SnapshotTesting
import SwiftUI
import Testing
@testable import LaneShadow

@Suite("Idle/Planning Screen Variant Tests")
@MainActor
struct IdlePlanningVariantTests {
    // MARK: - AC-1: Idle V01 no-location

    @Test("AC-1: Idle V01 no-location renders copper pill and dim chat input")
    func idleV01NoLocation() {
        // GIVEN: IdleScreen story "templates.idle-screen.v-no-location"
        let state = IdleMockProvider.value(variant: "v-no-location")

        // THEN: Location mode should be "needed" and greeting should reflect no location
        #expect(state.locationContext.mode == "needed", "Location mode should be 'needed'")
        #expect(state.greeting.meta.contains("Set a start"), "Meta should prompt to set start")
        #expect(state.greeting.headline.contains("starting"), "Headline should ask about starting point")

        // Verify UI renders correctly with copper pill and dim chat input
        let idleScreen = IdleScreen(provider: IdleMockProvider.self, variant: "v-no-location")
        let themedView = idleScreen.laneShadowTheme()

        let hostingController = UIHostingController(rootView: themedView)
        hostingController.view.frame = CGRect(x: 0, y: 0, width: 390, height: 844)
        hostingController.loadViewIfNeeded()

        // Verify the idle screen renders successfully
        #expect(hostingController.view != nil, "IdleScreen V01 should render successfully")

        // Verify via snapshot that copper pill and dim chat input are present
        assertSnapshot(
            matching: themedView,
            as: .image(precision: 0.9, traits: UITraitCollection(traitsFrom: [
                UITraitCollection(userInterfaceStyle: .light),
                UITraitCollection(userInterfaceIdiom: .phone),
                UITraitCollection(horizontalSizeClass: .compact),
                UITraitCollection(verticalSizeClass: .regular),
            ]))
        )
    }

    // MARK: - AC-2: Idle V02 first-ride

    @Test("AC-2: Idle V02 first-ride has no favorite pins and onboarding chips")
    func idleV02FirstRide() {
        // GIVEN: IdleScreen story "templates.idle-screen.v-first-ride"
        let state = IdleMockProvider.value(variant: "v-first-ride")

        // THEN: Suggestions should be onboarding-focused, not ride-specific
        #expect(!state.suggestions.isEmpty, "V02 should have onboarding suggestions")
        let onboardingChips = state.suggestions.map(\.label)
        #expect(onboardingChips.contains("Short & scenic") || onboardingChips.contains("Learn the roads"),
                "V02 should include onboarding chips like 'Short & scenic' or 'Learn the roads'")

        // Verify UI renders correctly with onboarding chips (no favorite pins)
        let idleScreen = IdleScreen(provider: IdleMockProvider.self, variant: "v-first-ride")
        let themedView = idleScreen.laneShadowTheme()

        let hostingController = UIHostingController(rootView: themedView)
        hostingController.view.frame = CGRect(x: 0, y: 0, width: 390, height: 844)
        hostingController.loadViewIfNeeded()

        // Verify the idle screen renders successfully
        #expect(hostingController.view != nil, "IdleScreen V02 should render successfully")

        // Verify via snapshot that onboarding chips are present and no favorite pins
        assertSnapshot(
            matching: themedView,
            as: .image(precision: 0.9, traits: UITraitCollection(traitsFrom: [
                UITraitCollection(userInterfaceStyle: .light),
                UITraitCollection(userInterfaceIdiom: .phone),
                UITraitCollection(horizontalSizeClass: .compact),
                UITraitCollection(verticalSizeClass: .regular),
            ]))
        )
    }

    // MARK: - AC-3: Idle V03 weather-advisory

    @Test("AC-3: Idle V03 weather-advisory renders warning meta and advisory card")
    func idleV03WeatherAdvisory() {
        // GIVEN: IdleScreen story "templates.idle-screen.v-weather-advisory"
        let state = IdleMockProvider.value(variant: "v-weather-advisory")

        // THEN: Greeting should reflect weather concern and meta should indicate severity
        #expect(state.greeting.meta.contains("Heavy rain") || state.greeting.meta.contains("Rain"),
                "V03 meta should mention rain/advisory")
        #expect(!state.greeting.headline.isEmpty, "V03 should have weather-concerned headline")

        // Verify UI renders correctly with warning meta and advisory card
        let idleScreen = IdleScreen(provider: IdleMockProvider.self, variant: "v-weather-advisory")
        let themedView = idleScreen.laneShadowTheme()

        let hostingController = UIHostingController(rootView: themedView)
        hostingController.view.frame = CGRect(x: 0, y: 0, width: 390, height: 844)
        hostingController.loadViewIfNeeded()

        // Verify the idle screen renders successfully
        #expect(hostingController.view != nil, "IdleScreen V03 should render successfully")

        // Verify via snapshot that warning meta and advisory card are present
        assertSnapshot(
            matching: themedView,
            as: .image(precision: 0.9, traits: UITraitCollection(traitsFrom: [
                UITraitCollection(userInterfaceStyle: .light),
                UITraitCollection(userInterfaceIdiom: .phone),
                UITraitCollection(horizontalSizeClass: .compact),
                UITraitCollection(verticalSizeClass: .regular),
            ]))
        )
    }

    // MARK: - AC-4: Planning V01 slow-planning

    @Test("AC-4: Planning V01 slow-planning renders italic apology")
    func planningV01Slow() {
        // GIVEN: PlanningScreen story "templates.planning-screen.v-slow"
        let state = PlanningMockProvider.value(variant: "v-slow")

        // THEN: Message should contain slow-planning apology
        #expect(state.message.body.contains("Still scouting") ||
            state.message.body.contains("taking longer") ||
            state.message.detail?.contains("scouting") == true,
            "V01 should have slow-planning apology message")

        // Verify UI renders correctly with italic apology note
        let planningScreen = PlanningScreen(provider: PlanningMockProvider.self, variant: "v-slow")
        let themedView = planningScreen.laneShadowTheme()

        let hostingController = UIHostingController(rootView: themedView)
        hostingController.view.frame = CGRect(x: 0, y: 0, width: 390, height: 844)
        hostingController.loadViewIfNeeded()

        // Verify the planning screen renders successfully
        #expect(hostingController.view != nil, "PlanningScreen V01 should render successfully")

        // Verify via snapshot that italic apology note is present
        assertSnapshot(
            matching: themedView,
            as: .image(precision: 0.9, traits: UITraitCollection(traitsFrom: [
                UITraitCollection(userInterfaceStyle: .light),
                UITraitCollection(userInterfaceIdiom: .phone),
                UITraitCollection(horizontalSizeClass: .compact),
                UITraitCollection(verticalSizeClass: .regular),
            ]))
        )
    }

    // MARK: - AC-5: Planning V02 cancel-confirm

    @Test("AC-5: Planning V02 cancel-confirm renders confirm sheet")
    func planningV02CancelConfirm() {
        // GIVEN: PlanningScreen story "templates.planning-screen.v-cancel-confirm"
        let state = PlanningMockProvider.value(variant: "v-cancel-confirm")

        // THEN: State should indicate cancel-confirm is active
        #expect(state.showCancelConfirm == true, "V02 should have cancel-confirm active")
        #expect(!state.phases.isEmpty, "V02 should have planning phases")

        // Verify UI renders correctly with cancel confirm sheet overlay
        let planningScreen = PlanningScreen(provider: PlanningMockProvider.self, variant: "v-cancel-confirm")
        let themedView = planningScreen.laneShadowTheme()

        let hostingController = UIHostingController(rootView: themedView)
        hostingController.view.frame = CGRect(x: 0, y: 0, width: 390, height: 844)
        hostingController.loadViewIfNeeded()

        // Verify the planning screen renders successfully with cancel sheet
        #expect(hostingController.view != nil, "PlanningScreen V02 should render successfully")

        // Verify via snapshot that cancel confirm sheet is rendered
        assertSnapshot(
            matching: themedView,
            as: .image(precision: 0.9, traits: UITraitCollection(traitsFrom: [
                UITraitCollection(userInterfaceStyle: .light),
                UITraitCollection(userInterfaceIdiom: .phone),
                UITraitCollection(horizontalSizeClass: .compact),
                UITraitCollection(verticalSizeClass: .regular),
            ]))
        )
    }

    // MARK: - AC-6: Planning V03 single-candidate

    @Test("AC-6: Planning V03 single-candidate renders warning chrome")
    func planningV03SingleCandidate() {
        // GIVEN: PlanningScreen story "templates.planning-screen.v-single-candidate"
        let state = PlanningMockProvider.value(variant: "v-single-candidate")

        // THEN: State should reflect single-candidate constraint
        #expect(!state.phases.isEmpty, "V03 should have planning phases")

        // Verify UI renders correctly with warning chrome
        let planningScreen = PlanningScreen(provider: PlanningMockProvider.self, variant: "v-single-candidate")
        let themedView = planningScreen.laneShadowTheme()

        let hostingController = UIHostingController(rootView: themedView)
        hostingController.view.frame = CGRect(x: 0, y: 0, width: 390, height: 844)
        hostingController.loadViewIfNeeded()

        // Verify the planning screen renders successfully
        #expect(hostingController.view != nil, "PlanningScreen V03 should render successfully")

        // Verify via snapshot that warning chrome is rendered
        assertSnapshot(
            matching: themedView,
            as: .image(precision: 0.9, traits: UITraitCollection(traitsFrom: [
                UITraitCollection(userInterfaceStyle: .light),
                UITraitCollection(userInterfaceIdiom: .phone),
                UITraitCollection(horizontalSizeClass: .compact),
                UITraitCollection(verticalSizeClass: .regular),
            ]))
        )
    }
}

File: android/app/src/testDebug/java/com/laneshadow/sandbox/IdlePlanningVariantTests.kt
package com.laneshadow.sandbox

import com.laneshadow.sandbox.mockproviders.IdleMockProvider
import com.laneshadow.sandbox.mockproviders.PlanningMockProvider
import org.junit.Test
import org.junit.Assert.*

/**
 * TDD tests for Idle and Planning screen variants (FID-S02-T04).
 *
 * Tests the six new edge-state variants:
 * - Idle V01: no-location (copper pill + dim chat input)
 * - Idle V02: first-ride (no pins + onboarding chips)
 * - Idle V03: weather-advisory (warning meta + advisory card)
 * - Planning V01: slow (italic apology + dashed border)
 * - Planning V02: cancel-confirm (dimmed card + scrim + confirm sheet)
 * - Planning V03: single-candidate (warning border + compass chip + header string)
 */
class IdlePlanningVariantTests {

    // ========================================================================
    // AC-1: Idle V01 no-location
    // ========================================================================

    @Test
    fun testIdleV01NoLocation() {
        // GIVEN: Android sandbox story templates.idle-screen.v-no-location is rendered
        val state = IdleMockProvider.value("v-no-location")

        // THEN: Location context shows "Tap to set start"
        assertEquals("Tap to set start", state.locationContext.label)

        // THEN: isNoLocation flag is true
        assertTrue(state.isNoLocation)

        // THEN: Advisory card is not shown
        assertFalse(state.showAdvisoryCard)
    }

    // ========================================================================
    // AC-2: Idle V02 first-ride
    // ========================================================================

    @Test
    fun testIdleV02FirstRide() {
        // GIVEN: Android sandbox story templates.idle-screen.v-first-ride is rendered
        val state = IdleMockProvider.value("v-first-ride")

        // THEN: Onboarding chips are present
        assertEquals(2, state.suggestions.size)
        assertTrue(state.suggestions.any { it.label == "Short & scenic" })
        assertTrue(state.suggestions.any { it.label == "Learn the roads" })

        // THEN: Advisory card is not shown
        assertFalse(state.showAdvisoryCard)

        // THEN: isNoLocation is false
        assertFalse(state.isNoLocation)
    }

    // ========================================================================
    // AC-3: Idle V03 weather-advisory
    // ========================================================================

    @Test
    fun testIdleV03WeatherAdvisory() {
        // GIVEN: Android sandbox story templates.idle-screen.v-weather-advisory is rendered
        val state = IdleMockProvider.value("v-weather-advisory")

        // THEN: Advisory card should be shown
        assertTrue(state.showAdvisoryCard)

        // THEN: Advisory message is present
        assertEquals("Rain expected", state.advisoryMessage)

        // THEN: Meta text indicates rain
        assertTrue(state.greeting.meta.contains("RAIN EXPECTED"))
    }

    // ========================================================================
    // AC-4: Planning V01 slow
    // ========================================================================

    @Test
    fun testPlanningV01Slow() {
        // GIVEN: Android sandbox story templates.planning-screen.slow-planning-light (or -dark) is rendered
        val state = PlanningMockProvider.value("v-slow")

        // THEN: Slow apology message is present
        assertNotNull(state.slowApology)
        assertTrue(state.slowApology!!.isNotEmpty())

        // THEN: No cancel confirm
        assertFalse(state.showCancelConfirm)

        // THEN: No warning border
        assertFalse(state.warningBorder)
    }

    // ========================================================================
    // AC-5: Planning V02 cancel-confirm
    // ========================================================================

    @Test
    fun testPlanningV02CancelConfirm() {
        // GIVEN: Android sandbox story templates.planning-screen.cancel-prompt-light (or -dark) is rendered
        val state = PlanningMockProvider.value("v-cancel-confirm")

        // THEN: Show cancel confirm flag is true
        assertTrue(state.showCancelConfirm)

        // THEN: No slow apology
        assertNull(state.slowApology)

        // THEN: No warning border
        assertFalse(state.warningBorder)
    }

    // ========================================================================
    // AC-6: Planning V03 single-candidate + header string
    // ========================================================================

    @Test
    fun testPlanningV03SingleCandidateAndHeader() {
        // GIVEN: Android sandbox story templates.planning-screen.single-candidate-light (or -dark) is rendered
        val state = PlanningMockProvider.value("v-single-candidate")

        // THEN: Warning border flag is true
        assertTrue(state.warningBorder)

        // THEN: Phase headers map is present
        assertNotNull(state.phaseHeaders)
        assertTrue(state.phaseHeaders!!.isNotEmpty())

        // THEN: No cancel confirm
        assertFalse(state.showCancelConfirm)
    }
}
